exec/check-lockfiles-up-to-date/src/checkLockfilesUpToDate.ts

```diff
@@ -0,0 +1,387 @@
+import filter from 'ramda/src/filter'
+import once from 'ramda/src/once'
+import { type Config, type OptionsFromRootManifest, getOptionsFromRootManifest } from '@pnpm/config'
+import { MANIFEST_BASE_NAMES, WANTED_LOCKFILE } from '@pnpm/constants'
+import { hashObjectNullableWithPrefix } from '@pnpm/crypto.object-hasher'
+import { PnpmError } from '@pnpm/error'
+import {
+  type Lockfile,
+  getLockfileImporterId,
+  readCurrentLockfile,
+  readWantedLockfile,
+} from '@pnpm/lockfile.fs'
+import {
+  calcPatchHashes,
+  createOverridesMapFromParsed,
+  getOutdatedLockfileSetting,
+} from '@pnpm/lockfile.settings-checker'
+import {
+  linkedPackagesAreUpToDate,
+  getWorkspacePackagesByDirectory,
+  satisfiesPackageManifest,
+} from '@pnpm/lockfile.verification'
+import { globalWarn, logger } from '@pnpm/logger'
+import { arrayOfWorkspacePackagesToMap } from '@pnpm/get-context'
+import { parseOverrides } from '@pnpm/parse-overrides'
+import { type WorkspacePackages } from '@pnpm/resolver-base'
+import {
+  type DependencyManifest,
+  type Project,
+  type ProjectId,
+  type ProjectManifest,
+} from '@pnpm/types'
+import { findWorkspacePackages } from '@pnpm/workspace.find-packages'
+import { loadPackagesList, updatePackagesList } from '@pnpm/workspace.packages-list-cache'
+import { readWorkspaceManifest } from '@pnpm/workspace.read-manifest'
+import { assertLockfilesEqual } from './assertLockfilesEqual'
+import { statManifestFile } from './statManifestFile'
+
+export type CheckLockfilesUpToDateOptions = Pick<Config,
+| 'allProjects'
+| 'autoInstallPeers'
+| 'cacheDir'
+| 'catalogs'
+| 'excludeLinksFromLockfile'
+| 'linkWorkspacePackages'
+| 'hooks'
+| 'peersSuffixMaxLength'
+| 'rootProjectManifest'
+| 'rootProjectManifestDir'
+| 'sharedWorkspaceLockfile'
+| 'virtualStoreDir'
+| 'workspaceDir'
+>
+
+export async function checkLockfilesUpToDate (opts: CheckLockfilesUpToDateOptions): Promise<void> {
+  const {
+    allProjects,
+    autoInstallPeers,
+    cacheDir,
+    catalogs,
+    excludeLinksFromLockfile,
+    linkWorkspacePackages,
+    rootProjectManifest,
+    rootProjectManifestDir,
+    sharedWorkspaceLockfile,
+    virtualStoreDir,
+    workspaceDir,
+  } = opts
+
+  if (!virtualStoreDir) return
+
+  const rootManifestOptions = rootProjectManifest && rootProjectManifestDir
+    ? getOptionsFromRootManifest(rootProjectManifestDir, rootProjectManifest)
+    : undefined
+
+  if (allProjects && workspaceDir) {
+    const packagesList = await loadPackagesList({ cacheDir, workspaceDir })
+    if (!packagesList) {
+      throw new PnpmError('RUN_CHECK_DEPS_NO_CACHE', 'Cannot check whether dependencies are outdated', {
+        hint: 'Run `pnpm install` to create the cache',
+      })
+    }
+
+    if (!equals(
+      filter(value => !!value, packagesList.catalogs ?? {}),
+      filter(value => !!value, catalogs ?? {})
+    )) {
+      throw new PnpmError('RUN_CHECK_DEPS_OUTDATED', 'Catalogs cache outdated', {
```

Comment:
why does the catalogs check happen before the modified projects check? Can’t we just check when `pnpm-workspace.yaml` was modified? If it wasn’t modified then probably catalogs didn’t change. Also, in that case I don’t see the list of catalogs should be cached. The catalogs can be checked later from the current lockfile.

Reply:
The list of catalogs is already stored inside "packages list" (this term may change), which mean it is already loaded along side "packages list". The "packages list" is already written to disk on `pnpm install` where `opts.catalogs` is already loaded.

In `pnpm run`, I don't think there is anyway to avoid loading the "packages list" as I always need the `lastValidatedTimestamp` and the list of `projectRootDirs` from the last `pnpm install` to compare. So if we can't avoid having to load the "packages list", might as well load catalogs from the same file.

I don't think invoking a syscall of `stat` is significantly more efficient than doing some in-memory object manipulation without the `stat` syscall.